### ISSUE TITLE
Clarification about the handling of sparse fieldset in documentation

### ISSUE THREAD

**Issue by krainboltgreene** [json-api/json-api#1305]
Given this request/response:

```
GET /accounts
{
	"links": {
		"self": "http://localhost:3001/accounts",
		"next": "http://localhost:3001/accounts?page[offset]=2&page[per]=10",
		"last": "http://localhost:3001/accounts?page[offset]=101&page[per]=10"
	},
	"data": [
		{
			"id": "cdb211e6-8eaa-44a3-9cad-e63da0b29a4e",
			"type": "accounts",
			"links": {
				"self": "http://localhost:3001/accounts/cdb211e6-8eaa-44a3-9cad-e63da0b29a4e"
			}
		}
	]
}
```

The documentation doesn't make it clear if I can then confidantly request:

```
GET /accounts?fields[accounts]=email
```

The current documentation [makes no suggestion about what to do in this situation](http://jsonapi.org/format/#fetching-sparse-fieldsets). Do I return nothing? Do I return a 422? Do I return the full set of fields?

Further, if my default payload looks like this:

```
GET /accounts
{
	"links": {
		"self": "http://localhost:3001/accounts",
		"next": "http://localhost:3001/accounts?page[offset]=2&page[per]=10",
		"last": "http://localhost:3001/accounts?page[offset]=101&page[per]=10"
	},
	"data": [
		{
			"id": "cdb211e6-8eaa-44a3-9cad-e63da0b29a4e",
			"type": "accounts",
			"attributes": {
				"name": "John Arbuckle",
				"age": 23
			}
			"links": {
				"self": "http://localhost:3001/accounts/cdb211e6-8eaa-44a3-9cad-e63da0b29a4e"
			}
		}
	]
}
```

Can I safely assume I can make this request:

```
GET /accounts?fields[accounts]=age
```

While the default payload contains it, there's nothing in the spec that guarantees my ability to specify it.

**Comment by krainboltgreene** [comment]
I thought the lack of a failure a particularly strange omission since `include` has this:

> If an endpoint does not support the include parameter, it MUST respond with 400 Bad Request to any requests that include it.

**Comment by krainboltgreene** [comment]
I also just noticed that `include` also has a passage about an include chain HAVING to be a relationship:

> The value of the include parameter MUST be a comma-separated (U+002C COMMA, “,”) list of relationship paths. **A relationship path is a dot-separated (...) list of relationship names.**

**Comment by ethanresnick** [comment]
> `GET /accounts?fields[accounts]=email`
I agree that the spec is unclear on how to handle a request like this, and I'm open to a PR that clarifies it. I think there are two options: 

1. return empty resource objects. In other words, you're treating the request as valid. Then, sparse fieldsets says that you can't return any fields except those specified, but the resources have no email, so they end up empty.

2. return a 400 error, like we do with the `?include` when we say: "If a server is unable to identify a relationship path... it MUST respond with 400 Bad Request." In other words, if the server *knows* this field name isn't valid, it's an error to send it.

In the abstract, my strong preference would be for the second option. (It catches a bug earlier, makes it clearer to the client what went wrong, and stops the client's response from magically changing at some point if the API does add the field.) In practice, though, I don't know how aggressive we can be in mandating something new, so I think we should look at how existing implementations handle this case and guide our behavior from that.

**Comment by krainboltgreene** [comment]
I would absolutely agree with matching includes in every way, but it also presents what many would consider breaking change (undefined behavior -> extreme behavior). How does the json-api committee want to handle that?
